WIP: vmware_guest_facts: Add lookup based on Manage Object ID (MoID)

wilmardo opened the pull request:
##### SUMMARY
Adds lookup based on Manage Object ID (moRef ID or moId) to vmware_guest_facts.
The moRef ID has been a unique identifier for years, see for example this good serie of @lamw about this:
https://www.virtuallyghetto.com/2017/07/uniquely-identifying-vms-in-vsphere-part-3-enhanced-linked-mode-cross-vc-vmotion.html

At Intermax all our tools are based on this identifier, I think we are not the only one with this 'problem'. This PR allows users to use this existing reference to gather the UUID or name/folder for future usage.

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Feature Pull Request

##### COMPONENT NAME
vmware_guest_facts

##### ADDITIONAL INFORMATION

Example of the usage in one of our roles:
```yaml
---
- hosts: localhost
  gather_facts: false
  tasks:
  - name: Gather facts for VMWare guest with Manage Object ID
    vmware_guest_facts:
      hostname: "{{ vmware_vcenter_hostname }}"
      username: "{{ vmware_vcenter_username }}"
      password: "{{ vmware_vcenter_password }}"
      datacenter: "{{ vmware_vcenter_datacenter }}"
      validate_certs: "{{ vmware_vcenter_validate_certs }}"
      moid: vm-31352
    register: facts

  - name: Use this new information to gather info again
    vmware_guest_facts:
      hostname: "{{ vmware_vcenter_hostname }}"
      username: "{{ vmware_vcenter_username }}"
      password: "{{ vmware_vcenter_password }}"
      datacenter: "{{ vmware_vcenter_datacenter }}"
      validate_certs: "{{ vmware_vcenter_validate_certs }}"
      name: "{{ facts.instance.hw_name }}"
      folder: "{{ facts.instance.hw_folder }}"
```

wilmardo committed:
Add lookup based on moid to vmware_guest_facts

jeking3 commented on lib/ansible/modules/cloud/vmware/vmware_guest_facts.py:
```diff
@@ -44,6 +44,11 @@
+     description:
+     - Manage Object ID of the instance to manage if known, this is an unique identifier only within a single vCenter instance.
+     - This is required if name or UUID is not supplied.
+     version_added: '2.8'
```
Probably needs to be changed to 2.9 now.

jeking3 commented on lib/ansible/modules/cloud/vmware/vmware_guest_facts.py:
```diff
@@ -44,6 +44,11 @@
      description:
      - UUID of the instance to manage if known, this is VMware's unique identifier.
      - This is required if name is not supplied.
```
if name or moid is not...
and change line 37

wilmardo replied:
Thanks for the catch, updated!

jeking3 commented on lib/ansible/module_utils/vmware.py:
```diff
@@ -952,6 +952,11 @@ def get_vm(self):
                 # Unique virtual machine found.
                 vm_obj = vms[0]
+        elif self.params['moid']:
+            objects = self.content.viewManager.CreateContainerView(self.content.rootFolder, [vim.VirtualMachine], True).view
```
More efficient way is to make an object with the moId.  Try:
```
    vm_obj = VmomiSupport.templateOf('VirtualMachine')(self.params['moid'], si._stub)
```
Must faster and doesn't involve a search on the server.

wilmardo replied:
Nice one, I am not all familair with pyvmomi so massive thanks for the suggestion!

jeking3 commented on lib/ansible/modules/cloud/vmware/vmware_guest_facts.py:
```diff
@@ -44,6 +44,11 @@
      - This is required if name is not supplied.
+   moid:
+     description:
+     - Manage Object ID of the instance to manage if known, this is an unique identifier only within a single vCenter instance.
```
```suggestion
     - Managed Object ID of the instance to manage if known, this is a unique identifier only within a single vCenter instance.
```

wilmardo replied:
Thanks again! Had to look up why `a unique` and not `an unique`, learned the rule that a vowel is an but there are some exceptions. You learn every day :)

jeking3 reviewed (changes requested):
Recommend a few changes before merging.

wilmardo commented:
@jeking3 Thanks for the review! Updated the PR with the changes, also rebased against the latest devel and squashed into one commit.

ready_for_review 🎉

wilmardo commented:
When #53523 is merged I can write some tests, for now this isn't straightforward since the MoID isn't returned as a fact.

jeking3 commented on lib/ansible/module_utils/vmware.py:
```diff
@@ -953,6 +953,8 @@ def get_vm(self):
                 # Unique virtual machine found.
                 vm_obj = vms[0]
+        elif self.params['moid']:
+            vm_obj = VmomiSupport.templateOf('VirtualMachine')(self.params['moid'], si._stub)
```
In this context we need to replace `si` with the pyVmomi service instance, likely in the base class of this module.  

The sample I wrote that uses this line of code for efficient lookup by moid can be found here:  https://github.com/vmware/pyvmomi/blob/master/sample/dumpjson.py#L68

jeking3 commented:
#53523 was merged.

wilmardo commented:
@jeking3 Added testcases! 

Also I needed to rework vmware.py somewhat to be able to use the serviceInstance (connect_to_api now returns the serviceInstance instead of the content). This led to reworking the tests as well to retrieve the content.

In `vmware_deploy_ovf.py` I saw that the variable was named as `self.si` but it contained the content so renamed it to `self.content` while I was at it :)

jeking3 reviewed (approved):
Looks good.

wilmardo commented:
@jeking3 Rebased against latest devel :)

samdoran commented:
Failure in unit test `test.units.module_utils.test_vmware::test_vmdk_disk_path_split_negative`
```
mocker = <pytest_mock.MockFixture object at 0x3a909d0>
fake_ansible_module = <units.module_utils.test_vmware.FakeAnsibleModule object at 0x3a90850>

    def test_vmdk_disk_path_split_negative(mocker, fake_ansible_module):
        """ Test vmdk_disk_path_split function"""
        fake_ansible_module.params = test_data[0][0]
    
        mocker.patch('ansible.module_utils.vmware.connect_to_api', new=fake_connect_to_api)
        with pytest.raises(FailJson) as exec_info:
>           pyv = PyVmomi(fake_ansible_module)

test/units/module_utils/test_vmware.py:153: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <ansible.module_utils.vmware.PyVmomi object at 0x3a902d0>
module = <units.module_utils.test_vmware.FakeAnsibleModule object at 0x3a90850>

    def __init__(self, module):
        """
            Constructor
            """
        if not HAS_REQUESTS:
            module.fail_json(msg=missing_required_lib('requests'),
                             exception=REQUESTS_IMP_ERR)
    
        if not HAS_PYVMOMI:
            module.fail_json(msg=missing_required_lib('PyVmomi'),
                             exception=PYVMOMI_IMP_ERR)
    
        self.module = module
        self.params = module.params
        self.si = connect_to_api(self.module)
        self.current_vm_obj = None
        # Uses if statement since connect_to_api could return None
>       self.content = self.si.RetrieveContent() if self.si else None
E       AttributeError: type object 'MyContent' has no attribute 'RetrieveContent'

lib/ansible/module_utils/vmware.py:810: AttributeError
```

Akasurde commented:
@wilmardo Thanks for this PR. I created a new PR https://github.com/ansible/ansible/pull/59143 which supersedes this PR. Thanks.
